### ISSUE TITLE
Error when subscribing a generic `cls` parameter in a classmethod

### ISSUE THREAD

**Issue by Daraan** [microsoft/pyright#10708]
**Describe the bug**

I have a classmethod that should return a dynamic subclass with different type parameters. To be correct I would need to subscribe `class Inner(cls[X]): ...`, however this raises an error. It appears that in this context `cls` is partially treated as a `TypeVar`  - according to the error, guess as it is related to `type[Self]` - and hence an error is reported. As it is runtime valid and the return type is accepted this appears as a bug to me.

**Code or Screenshots**
Code sample in [pyright playground](https://pyright-play.net/?locale=en-us&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoF0SKJAT1y1IHNseAkQCCpLixYBjADbF69KKJD8A2gA0AugC4oAOkNTmchUoDyyNCE27jUB1AACpxRDRIAFmAAmLR1A%2BaMDY5NYAjGoAHloAFHL0ADQkqnoq6jFQAJRQALQAfJw8aGqWSNbRWnaszAEBrkoAkmEg8bL0lVl6UADEUAA8uVAAKsUAasQgUABE3LglAMpossBOZRUaa1YgVdPYShRE9AgARvTSIFi4VCeyaP51j4b69o9Q6CggpFDNpNYOfXMAGlXg4XPI3B5vH5ao4giE%2BH8QAAmSptJIpfhpVSVbJ5QpzErrGwxaoPR4NH4tdFdclvBzPUF1D6ob6-f69EhKNBRebSco%2BADc7zQ%2BEIACUPKzRvNjMT9IiIrF0rEslkWEA)

```python
from __future__ import annotations

from typing import Any

class Arg[X]: ...

class Outer[X]:

    @classmethod
    def inner1[x](cls, arg: Arg[x] ) -> type[Outer[x]]:

        class Inner(cls[x]):  # <- TypeVar "type[Self@Outer[X@Outer]]" is not subscriptable
            ...

        return Inner  # OK

    @classmethod
    def inner2[x](cls, arg: Arg[x] ) -> type[Outer[x]]:

        class Inner(cls):
            ...

        return Inner  # as expected; reportReturnType

Outer.inner1(Arg())
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

**Comment by Azureblade3808** [comment]
There are two critical problems with `inner1` in your code.

1.  The value of `cls` is not always `Outer` but can also be any subclass of `Outer`, therefore it is not always subscriptable.
2.  `x` is a type-var defined on `def inner1[x]`, but you are also using it to define `Inner` by `class Inner(cls[x])` where `x` will be used as a type-var of `Inner`. Such usage is considered invalid in the typing system.

**Comment by Daraan** [comment]
> There are two critical problems with `inner1` in your code.
> 
>     1. The value of `cls` is not always `Outer` but can also be any subclass of `Outer`, therefore it is not always subscriptable.

Thank you. Yes, in the meantime I realized that as well 😄, and had to check that first. A reword that `cls` *might* not be subscript able could bring clarity. 

>     2. `x` is a type-var defined on `def inner1[x]`, but you are also using it to define `Inner` by `class Inner(cls[x])` where `x` will be used as a type-var of `Inner`. Such usage is considered invalid in the typing system.

Hmm, so a change of the error would be more appropriate here?
From my current understanding I doubt it, but is there are valid way to forward the `x` of `Arg[x]` to the new class? I replaced it with `Any` instead of the `TypeVar` at the loss of some type completions :/

**Comment by Azureblade3808** [comment]
> From my current understanding I doubt it, but is there are valid way to forward the `x` of `Arg[x]` to the new class? I replaced it with `Any` instead of the `TypeVar` at the loss of some type completions :/

I think you are doing right here.

There is no way to determine what a type-var really means in runtime. So `class Inner(Outer[Any])` may be the best option.

**Comment by Glinte** [comment]
I haven't actually ran this code, but what is wrong with just subclassing `Outer` instead of `cls` in your methods?

```python
from __future__ import annotations

from typing import Any

class Arg[X]: ...

class Outer[X]:

    @classmethod
    def inner1[x](cls, arg: Arg[x] ) -> type[Outer[x]]:

        class Inner(Outer[x]):  # <- TypeVar "type[Self@Outer[X@Outer]]" is not subscriptable
            ...

        # no use case, but related:
        new_type = Outer[x]  # Outer[X@Outer] is already specialiezed 

        return Inner  # OK

    @classmethod
    def inner2[x](cls, arg: Arg[x] ) -> type[Outer[x]]:

        class Inner(Outer[x]):
            ...

        return Inner  # as expected; reportReturnType

Outer.inner1(Arg())
```
[playground](https://pyright-play.net/?locale=en-us&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoF0SKJAT1y1IHNseAkQCCpLixYBjADbF69KKJD8A2gA0AugC4oAOkNTmchUoDyyNCE27jUB1AACpxRDRIAFmAAmLR1A%2BaMDY5NYAjGoAHloAFHL0ADQkqnoq6jFQAJRQALQAfJw8aGqWSNbRWnaszAEBrkoAkmEgsWUVMVl6UADEUAA8uVAAKsUAasQgUABE3LglAMpossBO7TYaa1YgVdPYShRE9AgARvTSIFi4VCeyaP51j4b69o%2B9UBRQCPRoUNIKaGSJ2QUHQ8nKPh0D0epDQAHcYHNfgBeKDrSoOProzbrLT7EiydDEHxcKD0ebSLDEWRYNAALzQPigr0e6BQIFIUGasKm73MAGkWc4Gu4vL5oYFgqEeQAmSrxWRJFL8NKqDE5ApFealbaVaoS%2BryRRclptXWdKG1N4BZ5CgJs1Cc7nWTEkJRoKIUiEAblBaHwhAASh5HaN5sZ1vo%2BDzwrF0rEslkWEA)

**Comment by Daraan** [comment]
> I haven't actually ran this code, but what is wrong with just subclassing `Outer` instead of `cls` in your methods?
> 

Thank you very much for the input. But it should subclass subclasses. If `arg: Arg[x]` is still generic the resulting `Inner` should also be generic in the parameters that are left (in my setting I have 3 or 4).
I had dumb mistake in my thinking that using `x` forwards this 🙃, but I need to check the args of `arg` at runtime obviously if I am not satisfied with `Any`.

**Comment by Glinte** [comment]
I don't think it is correct for type checkers to assume your subclass is always subscriptable, it is possible that you have a specialized subclass. It seems to be working as expected that static type checkers have to assume `cls` is never subscriptable unless the class itself is marked as final, which seems to be another bug that pyright doesn't allow subscripting a final generic class. Although if `cls` is marked as `@final` it is trivial to annotate `cls` as the class itself.

```python
from __future__ import annotations

from typing import Any

class Arg[X]: ...

class Outer[X]:

    @classmethod
    def inner1[x](cls, arg: Arg[x]) -> type[Outer[x]]:

        class Inner(cls[x]):  # <- TypeVar "type[Self@Outer[X@Outer]]" is not subscriptable
            ...

        return Inner  # OK


class SpecializedOuter(Outer[int]):
    pass


# runtime: TypeError: <class '__main__.SpecializedOuter'> is not a generic class
SpecializedOuter.inner1(Arg[str]())
```

**Comment by Daraan** [comment]
I am going to close this.
I think another error message would have might it to me more clearly earlier what problems can occur, but that is a minor thing

Thank you for the comments and showing interest ❤

**Comment by Glinte** [comment]
To be clear I think pyright's diagnostic is wrong here because it seems to be assuming that cls is the parameterized version of the class. But it just happens that two wrongs make a right
